File: scripts/wcag_contrast_audit.js
/**
 * Audit WCAG AA de contraste sur toutes les pages publiques.
 * Injecte axe-core, filtre les violations "color-contrast" (AA 4.5:1 / 3:1).
 * Sortie : /app/scripts/wcag_contrast_report.json
 */
const { chromium } = require('playwright');
const fs = require('fs');
const path = require('path');

const BASE = 'https://consultation-astro.preview.emergentagent.com';

const PAGES = [
  { url: '/',                          label: 'Accueil' },
  { url: '/theme-natal',               label: 'Vente · Thème Natal' },
  { url: '/kabbale',                   label: 'Vente · Kabbale' },
  { url: '/astrocartographie',         label: 'Vente · Astrocartographie' },
  { url: '/karma-destin',              label: 'Vente · Karma & Destin' },
  { url: '/numerologie',               label: 'Vente · Numérologie' },
  { url: '/synastrie',                 label: 'Vente · Synastrie' },
  { url: '/livres',                    label: 'Livres · landing' },
  { url: '/nos-livres',                label: 'Nos livres · catalog' },
  { url: '/blog',                      label: 'Blog' },
  { url: '/temoignages',               label: 'Témoignages' },
  { url: '/contact',                   label: 'Contact' },
  { url: '/credits',                   label: 'Crédits · info' },
  { url: '/mentions-legales',          label: 'Mentions légales' },
  { url: '/cgv',                       label: 'CGV' },
  { url: '/services/tirage-tarot',     label: 'Tirage Tarot (gate)' },
  { url: '/services/chatia',           label: 'ChatIA (gate)' },
  { url: '/services/horoscope',        label: 'Horoscope' },
  { url: '/services/choix',            label: 'Choix (gate)' },
  { url: '/connexion',                 label: 'Connexion' },
  { url: '/inscription',               label: 'Inscription' },
];

const AXE_CDN = 'https://cdnjs.cloudflare.com/ajax/libs/axe-core/4.9.1/axe.min.js';

async function auditPage(page, url, label) {
  const violations = [];
  try {
    await page.goto(`${BASE}${url}`, { waitUntil: 'networkidle', timeout: 30000 });
    await page.waitForTimeout(1200);
    // Inject axe-core
    await page.addScriptTag({ url: AXE_CDN });
    await page.waitForFunction(() => window.axe, { timeout: 8000 });
    // Run only the color-contrast check
    const result = await page.evaluate(async () => {
      const r = await window.axe.run(document, {
        runOnly: { type: 'rule', values: ['color-contrast'] },
      });
      return r.violations;
    });
    for (const v of result) {
      for (const n of v.nodes) {
        // Extract text preview + colors
        const html = n.html || '';
        const contrast = n.any?.[0]?.data;
        violations.push({
          selector: n.target?.join(' > ') || '',
          html: html.slice(0, 200),
          fg: contrast?.fgColor || null,
          bg: contrast?.bgColor || null,
          ratio: contrast?.contrastRatio || null,
          expected: contrast?.expectedContrastRatio || null,
          fontSize: contrast?.fontSize || null,
          fontWeight: contrast?.fontWeight || null,
        });
      }
    }
  } catch (e) {
    return { url, label, error: e.message, violations: [] };
  }
  return { url, label, violations };
}

(async () => {
  const browser = await chromium.launch();
  const context = await browser.newContext({ viewport: { width: 1440, height: 900 } });
  const page = await context.newPage();

  const results = [];
  for (const p of PAGES) {
    process.stdout.write(`Auditing ${p.url}... `);
    const r = await auditPage(page, p.url, p.label);
    console.log(r.error ? `ERR (${r.error.slice(0, 60)})` : `${r.violations.length} violations`);
    results.push(r);
  }

  await browser.close();

  const outPath = '/app/scripts/wcag_contrast_report.json';
  fs.writeFileSync(outPath, JSON.stringify(results, null, 2));
  console.log(`\nReport → ${outPath}`);

  // Print concise summary
  console.log('\n══ SUMMARY ══');
  for (const r of results) {
    if (r.error) { console.log(`  ${r.label.padEnd(30)}  ERROR: ${r.error.slice(0, 50)}`); continue; }
    const flag = r.violations.length === 0 ? '✓' : '✗';
    console.log(`  ${flag} ${r.label.padEnd(30)}  ${r.violations.length} violations`);
  }
})();

File: scripts/wcag_contrast_audit.py
#!/usr/bin/env python3
"""Audit WCAG AA de contraste (couleur uniquement) sur les pages publiques.

Injecte axe-core via CDN, filtre la règle 'color-contrast' (seuils AA 4.5:1 / 3:1),
et écrit un rapport JSON + résumé texte lisible.
"""
import json
import sys
import time
from pathlib import Path
from playwright.sync_api import sync_playwright

BASE = 'https://consultation-astro.preview.emergentagent.com'
AXE_CDN = 'https://cdnjs.cloudflare.com/ajax/libs/axe-core/4.9.1/axe.min.js'

PAGES = [
    ('/', 'Accueil'),
    ('/theme-natal', 'Vente · Thème Natal'),
    ('/kabbale', 'Vente · Kabbale'),
    ('/astrocartographie', 'Vente · Astrocartographie'),
    ('/karma-destin', 'Vente · Karma & Destin'),
    ('/numerologie', 'Vente · Numérologie'),
    ('/synastrie', 'Vente · Synastrie'),
    ('/livres', 'Livres · landing'),
    ('/nos-livres', 'Nos livres · catalog'),
    ('/blog', 'Blog'),
    ('/temoignages', 'Témoignages'),
    ('/contact', 'Contact'),
    ('/credits', 'Crédits · info'),
    ('/mentions-legales', 'Mentions légales'),
    ('/cgv', 'CGV'),
    ('/services/tirage-tarot', 'Tirage Tarot (gate)'),
    ('/services/chatia', 'ChatIA (gate)'),
    ('/services/horoscope', 'Horoscope'),
    ('/services/choix', 'Choix (gate)'),
    ('/connexion', 'Connexion'),
    ('/inscription', 'Inscription'),
]

def audit_page(page, url, label):
    try:
        page.goto(f'{BASE}{url}', wait_until='networkidle', timeout=30000)
        page.wait_for_timeout(1200)
        page.add_script_tag(url=AXE_CDN)
        page.wait_for_function('window.axe !== undefined', timeout=8000)
        result = page.evaluate('''async () => {
            const r = await window.axe.run(document, {
                runOnly: { type: 'rule', values: ['color-contrast'] },
            });
            return r.violations;
        }''')
        violations = []
        for v in result:
            for n in v.get('nodes', []):
                data = (n.get('any') or [{}])[0].get('data', {}) or {}
                violations.append({
                    'selector': ' > '.join(n.get('target') or []),
                    'html': (n.get('html') or '')[:220],
                    'fg': data.get('fgColor'),
                    'bg': data.get('bgColor'),
                    'ratio': data.get('contrastRatio'),
                    'expected': data.get('expectedContrastRatio'),
                    'fontSize': data.get('fontSize'),
                    'fontWeight': data.get('fontWeight'),
                })
        return {'url': url, 'label': label, 'violations': violations, 'error': None}
    except Exception as e:
        return {'url': url, 'label': label, 'violations': [], 'error': str(e)[:200]}


def main():
    out = []
    with sync_playwright() as pw:
        browser = pw.chromium.launch(executable_path='/pw-browsers/chromium_headless_shell-1208/chrome-linux/headless_shell')
        ctx = browser.new_context(viewport={'width': 1440, 'height': 900})
        page = ctx.new_page()
        for u, l in PAGES:
            print(f'Auditing {u}... ', end='', flush=True)
            r = audit_page(page, u, l)
            if r['error']:
                print(f'ERR ({r["error"][:60]})')
            else:
                print(f'{len(r["violations"])} violations')
            out.append(r)
        browser.close()

    report_path = Path('/app/scripts/wcag_contrast_report.json')
    report_path.write_text(json.dumps(out, ensure_ascii=False, indent=2))
    print(f'\nReport → {report_path}')

    # Concise summary
    print('\n══ SUMMARY ══')
    total = 0
    for r in out:
        if r['error']:
            print(f'  ⚠ {r["label"]:<32} ERROR: {r["error"][:50]}')
            continue
        n = len(r['violations'])
        total += n
        flag = '✓' if n == 0 else '✗'
        print(f'  {flag} {r["label"]:<32} {n} violations')
    print(f'\nTotal contrast violations: {total}')


if __name__ == '__main__':
    main()
